### ISSUE TITLE
Chaining suspend functions causes a class cast exception in some cases

### ISSUE THREAD

**Issue by koush** [JetBrains/kotlin-native#3476]
I've been trying to get a standalone repro, but to no avail. The following code in fails with a ClassCastException (Unit is not UvSocket).

```kotlin
        val server = listen(0, null, 10000)
        connect("127.0.0.1", server.localPort).close()
```

This very same code simply rewritten as follows does not throw the exception:
```
        val server = listen(0, null, 10000)
        val s = connect("127.0.0.1", server.localPort)
        s.close()
```

UvSocket is roughly:
```kotlin
interface Socket {
  suspend fun close()
}
class UvSocket: Socket {
  // implementation
}
```

full stack:
```
kotlin.ClassCastException: kotlin.Unit cannot be cast to com.koushikdutta.scratch.event.UvSocket
        at 0   test.kexe                           0x0000000100001eb7 kfun:kotlin.Throwable.<init>(kotlin.String?)kotlin.Throwable + 23 (/Users/teamcity/buildAgent
/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/Throwable.kt:21:56)
        at 1   test.kexe                           0x0000000100001e75 kfun:kotlin.Exception.<init>(kotlin.String?)kotlin.Exception + 21 (/Users/teamcity/buildAgent
/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/Exceptions.kt:23:5)
        at 2   test.kexe                           0x0000000100001e35 kfun:kotlin.RuntimeException.<init>(kotlin.String?)kotlin.RuntimeException + 21 (/Users/teamc
ity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/Exceptions.kt:34:5)
        at 3   test.kexe                           0x000000010013cdc5 kfun:kotlin.ClassCastException.<init>(kotlin.String?)kotlin.ClassCastException + 21 (/Users/t
eamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/Exceptions.kt:106:5)
        at 4   test.kexe                           0x00000001001423fa ThrowClassCastException + 362 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/m
ain/kotlin/kotlin/native/internal/RuntimeUtils.kt:24:11)
        at 5   test.kexe                           0x0000000100159e13 CheckInstance + 83
        at 6   test.kexe                           0x0000000100110dc4 kfun:com.koushikdutta.scratch.LoopTests.$testSocketsALot$lambda-16COROUTINE$57.invokeSuspend#
internal + 1140 (/Volumes/Dev/scratch/src/commonTest/kotlin/com/koushikdutta/scratch/LoopTests.kt:<unknown>)
        at 7   test.kexe                           0x0000000100077965 kfun:kotlin.coroutines.native.internal.BaseContinuationImpl.resumeWith(kotlin.Result<kotlin.A
ny?>) + 325 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/coroutines/ContinuationImpl.kt:30:39)
        at 8   test.kexe                           0x0000000100099252 kfun:kotlin.coroutines.SafeContinuation.resumeWith(kotlin.Result<#GENERIC>) + 258 (/Users/tea
mcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/coroutines/SafeContinuationNative.kt:33:26)
        at 9   test.kexe                           0x00000001000c7423 kfun:com.koushikdutta.scratch.event.AsyncScheduler.post$<anonymous>_2#internal + 163 (/Users/
teamcity/buildAgent/work/4d622a065c544371/backend.native/build/stdlib/kotlin/coroutines/Continuation.kt:<unknown>)
        at 10  test.kexe                           0x00000001000c7375 kfun:com.koushikdutta.scratch.event.AsyncScheduler.$post$<anonymous>_2$FUNCTION_REFERENCE$67.
invoke#internal + 21 (/Volumes/Dev/scratch/src/commonMain/kotlin/com.koushikdutta.scratch/event/scheduler.kt:139:14)
        at 11  test.kexe                           0x00000001000c7345 kfun:com.koushikdutta.scratch.event.AsyncScheduler.$post$<anonymous>_2$FUNCTION_REFERENCE$67.$<bridge-UNN>invoke()#internal + 21 (/Volumes/Dev/scratch/src/commonMain/kotlin/com.koushikdutta.scratch/event/scheduler.kt:139:14)
        at 12  test.kexe                           0x0000000100108729 kfun:com.koushikdutta.scratch.event.AsyncScheduler.lockAndRunQueue()kotlin.Long + 441 (/Volumes/Dev/scratch/src/commonMain/kotlin/com.koushikdutta.scratch/event/scheduler.kt:72:19)
        at 13  test.kexe                           0x00000001001081f8 kfun:com.koushikdutta.scratch.event.UvEventLoop.run() + 488 (/Volumes/Dev/scratch/src/nativeMain/kotlin/com/koushikdutta/scratch/event/uvloop.kt:335:28)
        at 14  test.kexe                           0x0000000100107e87 kfun:com.koushikdutta.scratch.LoopTests.networkContextTest#internal + 247 (/Volumes/Dev/scratch/src/commonTest/kotlin/com/koushikdutta/scratch/LoopTests.kt:50:13)
        at 15  test.kexe                           0x0000000100107d51 kfun:com.koushikdutta.scratch.LoopTests.networkContextTest$default#internal + 113 (/Volumes/Dev/scratch/src/commonTest/kotlin/com/koushikdutta/scratch/LoopTests.kt:<unknown>)
        at 16  test.kexe                           0x00000001001108d8 kfun:com.koushikdutta.scratch.LoopTests.testSocketsALot() + 104 (/Volumes/Dev/scratch/src/commonTest/kotlin/com/koushikdutta/scratch/LoopTests.kt:<unknown>)
        at 17  test.kexe                           0x0000000100110868 kfun:com.koushikdutta.scratch.$LoopTests$test$0.$testSocketsALot$FUNCTION_REFERENCE$40.invoke#internal + 24 (/Volumes/Dev/scratch/src/commonTest/kotlin/com/koushikdutta/scratch/LoopTests.kt:185:6)
        at 18  test.kexe                           0x0000000100110839 kfun:com.koushikdutta.scratch.$LoopTests$test$0.$testSocketsALot$FUNCTION_REFERENCE$40.$<bridge-UNNN>invoke(#GENERIC)#internal + 25 (/Volumes/Dev/scratch/src/commonTest/kotlin/com/koushikdutta/scratch/LoopTests.kt:185:6)
        at 19  test.kexe                           0x00000001000e945d kfun:kotlin.native.internal.test.BaseClassSuite.TestCase.run() + 445 (/Users/teamcity/buildAgent/work/4d622a065c544371/backend.native/build/stdlib/generated/_Collections.kt:97:7)
        at 20  test.kexe                           0x0000000100010559 kfun:kotlin.native.internal.test.TestRunner.run#internal + 1001 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/native/internal/test/TestRunner.kt:223:0)
        at 21  test.kexe                           0x000000010000fcb1 kfun:kotlin.native.internal.test.TestRunner.runIteration#internal + 1217 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/system/Timing.kt:33:12)
        at 22  test.kexe                           0x000000010000f600 kfun:kotlin.native.internal.test.TestRunner.run()kotlin.Int + 544 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/native/internal/test/TestRunner.kt:270:0)
        at 23  test.kexe                           0x000000010000e7c2 kfun:kotlin.native.internal.test.testLauncherEntryPoint(kotlin.Array<kotlin.String>)kotlin.Int + 114 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/native/internal/test/Launcher.kt:19:47)
        at 24  test.kexe                           0x000000010000e741 kfun:kotlin.native.internal.test.main(kotlin.Array<kotlin.String>) + 17 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/main/kotlin/kotlin/native/internal/test/Launcher.kt:23:5)
        at 25  test.kexe                           0x000000010000e6bf Konan_start + 63 (/Users/teamcity/buildAgent/work/4d622a065c544371/runtime/src/launcher/kotlin/konan/start.kt:<unknown>)
        at 26  test.kexe                           0x000000010000e54e Init_and_run_start + 78
        at 27  libdyld.dylib                       0x00007fff5c9103d5 start + 1
```

**Comment by cy6erGn0m** [comment]
Looks like a wrong generated state machine, we had similar issues with JVM backend. 

@koush For sure we need a better reproducer for this. Could you please provide more code?

**Comment by koush** [comment]
Yes, I'll check in my entire repo, as I have intended it to be open source. It's an implementation of socket/http/http2 client and server in kotlin. It's using libuv in kotlin native.

There's a small test case that can reproduce the issue with consistency. I'll need to set up some quick build instructions as it requires building openssl and libuv locally for cinterop, and will update this bug.

**Comment by koush** [comment]
https://github.com/koush/scratch

**Comment by SvyatoslavScherbina** [comment]
Thank you for the detailed report, we've successfully reproduced the issue.

**Comment by koush** [comment]
Thanks. Realized I didn't point you to the failing tests, but I suspect you found them:

https://github.com/koush/scratch/blob/master/src/commonTest/kotlin/com/koushikdutta/scratch/KotlinBugs.kt

**Comment by SvyatoslavScherbina** [comment]
> testCapturedVariableIncrementInSuspend

This one is likely not a bug. See [the documentation on +=](https://kotlinlang.org/docs/reference/operator-overloading.html#assignments) for more details.

`count += connect("127.0.0.1", server.localPort).countBytes()` here is equivalent to
```kotlin
count = count + connect("127.0.0.1", server.localPort).countBytes()
```
i.e.

1.  Read `count` value
2.  Execute `connect(...).countBytes()`
3.  Assign the sum to `count` variable.

The expression in the second step can suspend, so it can interleave with other `async` coroutines.

So the following can actually happen:
1. Coroutine 1 reads `count` value (let it be `count1`) and suspends.
2. Coroutine 2 reads `count` value (it is still `count1`) and suspends.
3. Coroutine 1 resumes and assigns `count = count1 + delta1`.
4. Coroutine 2 resumes and assigns `count = count1 + delta2`.

So finally `count` is `count1 + delta2`, not `count1 + delta1 + delta2`.

**Comment by koush** [comment]
@SvyatoslavScherbina That seems odd, and I understand why it is done that way.

Inc/dec operators should probably try to resolve a rhs before reading the value. I understand these operators are never truly atomic, but the disconnect between the count's read and write ventures well into the unexpected behavior here.

**Comment by koush** [comment]
Coroutines aside, I'd also say this is a strange behavior of the language that would also reproduce this behavior:

Thread 1:
```
x += somethingThatTakesALongTime()
```

Thread 2:
```
doSomethingToChangeValueOfX()
```

Where Thread 1 reads the value of X, blocks for some time, and Thread 2 changes the value of x underneath.

I'm not familiar with how other languages have handled this, so maybe this is a common behavior.

**Comment by koush** [comment]
Messaged a friend on the C# team that implemented await @jaredpar, and he confirmed that Kotlin increment implementing behaves the same as C#.

**Comment by SvyatoslavScherbina** [comment]
`testChainingSuspendSocketClose` is fixed in master with #3526.
It will be included into 1.3.70, not sure about 1.3.60 though.

**Comment by Thomas-Vos** [comment]
I am experiencing this issue, too. Changing the code to multiple lines fixes the issue. I guess your fix didn't make it to 1.3.61?

**Comment by SvyatoslavScherbina** [comment]
> I guess your fix didn't make it to 1.3.61?

Correct.
